app/src/backlight.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <init.h>
+#include <kernel.h>
+
+#include <drivers/led.h>
+#include <logging/log.h>
+#include <settings/settings.h>
+
+#include <zmk/activity.h>
+#include <zmk/backlight.h>
+#include <zmk/usb.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/activity_state_changed.h>
+#include <zmk/events/usb_conn_state_changed.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+BUILD_ASSERT(DT_HAS_CHOSEN(zmk_backlight),
+             "CONFIG_ZMK_BACKLIGHT is enabled but no zmk,backlight chosen node found");
+
+static const struct device *const backlight_dev = DEVICE_DT_GET(DT_CHOSEN(zmk_backlight));
+
+#define CHILD_COUNT(...) +1
+#define DT_NUM_CHILD(node_id) (DT_FOREACH_CHILD(node_id, CHILD_COUNT))
+
+#define BACKLIGHT_NUM_LEDS (DT_NUM_CHILD(DT_CHOSEN(zmk_backlight)))
+
+#define BRT_MAX 100
+
+struct backlight_state {
+    uint8_t brightness;
+    bool on;
+};
+
+static struct backlight_state state = {.brightness = CONFIG_ZMK_BACKLIGHT_BRT_START,
+                                       .on = IS_ENABLED(CONFIG_ZMK_BACKLIGHT_ON_START)};
+
+static int zmk_backlight_update() {
+    uint8_t brt = zmk_backlight_get_brt();
+    LOG_DBG("Update backlight brightness: %d%%", brt);
+
+    for (int i = 0; i < BACKLIGHT_NUM_LEDS; i++) {
+        int rc = led_set_brightness(backlight_dev, i, brt);
+        if (rc != 0) {
+            return rc;
+        }
```

Comment:
Do we really want to exit the loop early if the update to one fails?

Perhaps better to `LOG_ERR` all failures, and return some generic errno if any failed.

Reply:
I can change it, however, if the update fails it's most likely due to these two reasons:

* The device is configured incorrectly (e.g. not a real LED PWM peripheral)
* The LED index is out of range (again, due to an incorrect configuration)

In the first case, every update fails, so it makes sense to stop at the first error.
In the second case, every update with an index out of range fails, again it makes sense to stop at the first error.

Reply:
WFM. Thanks.